File: sdk/python/agentfield/harness/_result.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional


@dataclass
class Metrics:
    duration_ms: int = 0
    duration_api_ms: int = 0
    num_turns: int = 0
    total_cost_usd: Optional[float] = None
    usage: Optional[Dict[str, Any]] = None
    session_id: str = ""


@dataclass
class RawResult:
    result: Optional[str] = None
    messages: List[Dict[str, Any]] = field(default_factory=list)
    metrics: Metrics = field(default_factory=Metrics)
    is_error: bool = False
    error_message: Optional[str] = None


@dataclass
class HarnessResult:
    result: Optional[str] = None
    parsed: Any = None
    is_error: bool = False
    error_message: Optional[str] = None
    cost_usd: Optional[float] = None
    num_turns: int = 0
    duration_ms: int = 0
    session_id: str = ""
    messages: List[Dict[str, Any]] = field(default_factory=list)

    @property
    def text(self) -> str:
        if self.result:
            return self.result
        return ""

File: sdk/go/harness/result.go
// Package harness dispatches tasks to external coding agents (CLI subprocesses)
// and extracts structured results with schema validation and retry logic.
package harness

// FailureType classifies how a harness invocation failed so the runner
// can decide on a retry strategy.
type FailureType string

const (
	FailureNone     FailureType = "none"
	FailureCrash    FailureType = "crash"
	FailureTimeout  FailureType = "timeout"
	FailureAPIError FailureType = "api_error"
	FailureNoOutput FailureType = "no_output"
	FailureSchema   FailureType = "schema"
)

// Metrics captured from a single provider execution.
type Metrics struct {
	DurationMS    int
	DurationAPIMS int
	NumTurns      int
	SessionID     string
}

// RawResult is the output from a single provider execution before schema
// parsing.
type RawResult struct {
	Result       string
	Messages     []map[string]any
	Metrics      Metrics
	IsError      bool
	ErrorMessage string
	FailureType  FailureType
	ReturnCode   int
}

// Result is the final harness output after schema validation, retries, and
// metrics accumulation.
type Result struct {
	// Result is the raw text output from the provider.
	Result string

	// Parsed holds the validated and deserialized structured output.
	// The caller passes a pointer to a struct; on success it is populated.
	Parsed any

	IsError      bool
	ErrorMessage string
	FailureType  FailureType

	NumTurns   int
	DurationMS int
	SessionID  string
	Messages   []map[string]any
}

// Text returns the result text, or empty string if nil.
func (r *Result) Text() string {
	return r.Result
}
